controllers/controller.go

```diff
@@ -15,3 +15,173 @@ limitations under the License.
+	// kubeClient communicates with Kubernetes API server
+	kubeClient kubernetes.Interface
+	// ociClient is the client to operates on ocibuilder resource
+	ociClient ociv1alpha1.Interface
+	// informer provides eventually consistent linkage of its clients to the authoritative state of a given collection of objects.
+	informer cache.SharedIndexInformer
+	// queue is an interface that rate limits items being added to the queue.
+	queue workqueue.RateLimitingInterface
+}
+
+// NewController creates a new controller
+func NewController(rest *rest.Config, config *ControllerConfig, logger *logrus.Logger, configmap, namespace string) *Controller {
+	rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(rateLimiterBaseDelay, rateLimiterMaxDelay)
+	return &Controller{
+		Namespace:  namespace,
+		Configmap:  configmap,
+		Config:     config,
+		logger:     logger,
+		kubeConfig: rest,
+		kubeClient: kubernetes.NewForConfigOrDie(rest),
+		ociClient:  ociv1alpha1.NewForConfigOrDie(rest),
+		queue:      workqueue.NewRateLimitingQueue(rateLimiter),
+	}
+}
+
+func (ctrl *Controller) processNextItem() bool {
+	// Wait until there is a new item in the queue
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	obj, exists, err := ctrl.informer.GetIndexer().GetByKey(key.(string))
+	if err != nil {
+		fmt.Printf("failed to get ocibuilder '%s' from informer index: %+v", key, err)
+		return true
+	}
+
+	if !exists {
+		// this happens after ocibuilder was deleted, but work queue still had entry in it
+		return true
+	}
+
+	builder, ok := obj.(*v1alpha1.OCIBuilder)
+	if !ok {
+		fmt.Printf("key '%s' in index is not a builder", key)
+		return true
+	}
+
+	ctx := newOperationContext(builder, ctrl)
+
+	err = ctx.operate()
+	if err != nil {
+		ctrl.logger.WithError(err).WithField(common.LabelOCIBuilderName, builder.Name).Errorln("failed to operate on the ocibuilder obejct")
+	}
+
+	err = ctrl.handleErr(err, key)
+	if err != nil {
+		ctrl.logger.WithError(err).Errorln("controller is unable to handle the error")
+	}
+	return true
+}
+
+// handleErr checks if an error happened and make sure we will retry later
+// returns an error if unable to handle the error
+func (ctrl *Controller) handleErr(err error, key interface{}) error {
+	if err == nil {
+		// Forget about the #AddRateLimited history of key on every successful sync
+		// Ensure future updates for this key are not delayed because of outdated error history
+		ctrl.queue.Forget(key)
+		return nil
+	}
+
+	// due to the base delay of 5ms of the DefaultControllerRateLimiter
+	// requeues will happen very quickly even after a ocibuilder pod goes down
+	// we want to give the ocibuilder pod a chance to come back up so we give a generous number of retries
+	if ctrl.queue.NumRequeues(key) < 20 {
+		// Re-enqueue the key rate limited. This key will be processed later again.
+		ctrl.queue.AddRateLimited(key)
+		return nil
+	}
+	return errors.New("exceeded max requeues")
```

Comment:
I think we should log the original error here as well to give the user some context on what has failed

Reply:
It is already logged before calling handleErr.